app-modules/meeting-center/src/Managers/OutlookCalendarManager.php

```diff
@@ -31,65 +31,281 @@
 
 use DateTime;
+use Carbon\Carbon;
+use DateTimeInterface;
+use Microsoft\Graph\Graph;
+use Microsoft\Graph\Model\Event;
+use Microsoft\Graph\Model\Attendee;
+use Microsoft\Graph\Model\BodyType;
+use Microsoft\Graph\Model\ItemBody;
+use Illuminate\Support\Facades\Http;
+use Microsoft\Graph\Model\EmailAddress;
+use Microsoft\Graph\Core\GraphConstants;
 use Assist\MeetingCenter\Models\Calendar;
+use GuzzleHttp\Exception\ClientException;
+use Microsoft\Graph\Model\DateTimeTimeZone;
 use Assist\MeetingCenter\Models\CalendarEvent;
 use Assist\MeetingCenter\Managers\Contracts\CalendarInterface;
 
 class OutlookCalendarManager implements CalendarInterface
 {
-    /**
-     * @return array<string, string>
-     */
     public function getCalendars(Calendar $calendar): array
     {
-        // TODO: Implement getCalendars() method.
+        $client = (new Graph())->setAccessToken($calendar->oauth_token);
 
-        return [];
-    }
+        $calendars = $client->createRequest('GET', '/me/calendars')
+            ->setReturnType(\Microsoft\Graph\Model\Calendar::class)
+            ->execute();
 
-    // https://github.com/microsoftgraph/msgraph-sample-phpapp/tree/main
-    // https://github.com/microsoftgraph/msgraph-sdk-php
-    // https://learn.microsoft.com/en-us/graph/api/resources/webhooks?view=graph-rest-1.0
+        return collect($calendars)->filter(fn (\Microsoft\Graph\Model\Calendar $item) => $item->getCanEdit())
+            ->mapWithKeys(fn (\Microsoft\Graph\Model\Calendar $item) => [$item->getId() => $item->getName()])
+            ->toArray();
+    }
 
     public function getEvents(Calendar $calendar, ?Datetime $start = null, ?Datetime $end = null, ?int $perPage = null): array
     {
-        // https://learn.microsoft.com/en-us/graph/api/user-list-calendars?view=graph-rest-1.0&tabs=http
-        // https://learn.microsoft.com/en-us/graph/api/user-list-events?view=graph-rest-1.0&tabs=http
+        $client = $this->makeClient($calendar);
+
+        $start = $start ?? now()->subYear()->startOfDay();
+
+        $end = $end ?? now()->addYear()->endOfDay();
+
+        $events = [];
+
+        $request = $client->createCollectionRequest(
+            requestType: 'GET',
+            endpoint: '/me/calendar/calendarView?' . http_build_query([
+                '$top' => $perPage ?? GraphConstants::MAX_PAGE_SIZE,
+                'startDateTime' => $start->format(DateTimeInterface::ATOM),
+                'endDateTime' => $end->format(DateTimeInterface::ATOM),
+            ])
+        );
 
-        // TODO: Implement getEvents() method.
+        do {
+            try {
+                $response = $request->execute();
+            } catch (ClientException $exception) {
+                if ($exception->getCode() === 401) {
+                    $calendar = $this->refreshToken($calendar);
 
-        return [];
+                    $request->setAccessToken($calendar->oauth_token);
+
+                    $response = $request->execute();
```

Comment:
You do this check throughout this file but is it necessary? In makeClient you already check and refresh the token so if it 401s again why try to refresh a second time?

If you do keep it what about `Symfony\Component\HttpFoundation\Response::HTTP_UNAUTHORIZED` instead of 401?

Reply:
It seems possible that the oauth token could expire after instantiating the client initially, so it probably makes sense to do this at some level. But, it might make sense for us to abstract this so it's done behind the scenes per request execution. That way we don't need to remember to do it in every instance.

Reply:
The check of whether or not we think the `access_token` is expired based on the time we stored it does happen in `makeClient`, but it is possible for something to go wrong with that instance of the access_token or Microsoft expires it for some reason. So these checks ensure that if we get a failure, we try and refresh.